docs/design/fabric-v2+/interfaces.ecc-tlcc-channel.md

```diff
@@ -0,0 +1,462 @@
+func tl_session_rpc(req bytes[]) (rsp bytes[], error)
+```
+
+
+### TLCC_Enclave
+
+```c++
+typedef struct tl_tlcc_ctx* tl_tlcc_ctx_ptr_t;
+
+// TLCC_Enclave provided handler for processing requests and providing corresponding responses
+// Notes:
+// - in case TLCC_Enclave handler doesn't have to do ocalls requiring the
+//	 untrusted shim context, that parameter could be dropped here (and
+//   below in `tl_session_rpc`..
+typedef int (*tl_request_handler_t)(
+	shim_ctx_ptr_t u_shim_ctx,
+	tl_tlcc_ctx_ptr_t ecc_ctx
+	uint8_t* request,
+	size_t request_len
+	uint8_t* response,
+	size_t* response_len,
+	size_t max_response_len);
+
+// register a request processor function
+int tl_session_register_request_handlers(
+	tl_request_handler_t handler_f);
+
+// return the channel_id associated with an (established) session
+int tl_session_get_channel_id(
+	tl_tlcc_ctx_ptr_t ecc_ctx,
+	char** channel_id,
+    size_t* channel_id_len,
+    size_t max_channel_id_len);
+
+// return the chaincode_id associated with an (established) session
+int tl_session_get_chaincode_id(
+	tl_tlcc_ctx_ptr_t ecc_ctx,
+	char** chaincode_id,
+    size_t* chaincode_id_len,
+    size_t max_chaincode_id_len);
+
+// return the enclave_id associated with an (established) session
+int tl_session_get_chaincode_id(
+	tl_tlcc_ctx_ptr_t ecc_ctx,
+	char* enclave_id);
+    size_t* enclave_id_len,
+    size_t max_enclave_id_len);
+```
+
+
+### Properties/Guarantees
+
+- after successful `tl_session_setup` any `tl_session_request` causes `req`
+  to be delivered to TLCC_Enclave in an _authentic_ (i.e. integrity protected) but _non
+  confidential_ manner; similarly, any response `response` is delivered in an
+  authentic manner as return parameter of `tl_session_request`.
+- for any request, TLCC_Enclave and ECC_Enclave agree on the set
+  `<chaincode_id, enclave_id, tlcc_mrenclave, channel_id, channel_hash>`
+  as returned by the `tl_session_get_*` functions and their view on their
+  own identity.
+- lastly, the module ensures that for the duration of a single
+  sessions, the interaction is between the same set of enclave
+  _instances_.
+
+## Internals
+
+### Session State
+
+#### ECC
+```c++
+    #include <sgx_dh.h>
+
+	// interal session context
+    enum tl_ecc_state {INIT1, INIT2, ACTIVE, CLOSING, CLOSED} tl_ecc_state_t;
+
+	struct tl_ecc_ctx {
+		session_id_t sid,
+    	tl_ecc_state_t state,
+		sgx_dh_session_t dh
+		sgx_key_128bit_t aek
+		char* cid,
+		channel_hash_t ch,
+		char* cc,
+		char* eid,
+		sgx_dh_session_enclave_identity_t tl_identity,
+	} tl_ecc_ctx_t;
+```
+
+
+#### TLCC_Enclave
+
+```c++
+  #include <sgx_dh.h>
+	#include <map>
+
+	// internal session context
+	// - ACTIVE means tl_session_setup succeeded
+	// - ACTIVE_VALIDATED means tl_session_setup succeeded _and_ we have verified cc & ecc_identity via ERCC
+    typedef tl_tlcc_state enum{INIT, ACTIVE, ACTIVE_VALIDATED, CLOSED} tl_tlcc_state_t;
```

Comment:
good question: regarding two states, conceptually you need it also to ensure that the view on `enclave_id` is consistent.  Regarding `chaincode_id`,  the setup protocol as is guarantees that ecc and tlcc enclaves do have agreement on it _if_ ecc sends a request.  So an adversary can not confuse tlcc of thinking a different chaincode applies than what an honest ecc thinks.  So if there is no request which requires any validation that ecc "belongs" to `chaincode_id`, then one could skip that state.  i've added it here to be on safe side.
=>
i think i would leave the states but if we think we do not need authenticated/validated `chaincode_id`, i would propose we either change `tl_session_get_chaincode_id` to `tl_session_get_unvalidated_chaincode_id` on the tlcc side or, simpler, just add a corresponding comment to `tl_session_get_chaincode_id` that it is _not_ (cross-)validated that enclave_id is indeed registered with it. The same also applies to any check that the mrenclave which is part of the local attestation matches the version registered with that chaincode.

Reply:
> One thing is unclear.
> Blocks and metadata might be updated in TLCC as blocks arrive, and the session does not ensure a "consistent view" (for a suitable definition of "consistent"), as it came up in #435 .
> In theory, this should not be a problem, because we can rely on the versioning of the ledger keys/values, which is checked at commit time. However, this is not straightforward.

I don't think the session module can do anything regarding this on its own.  This has to be handled on a higher legel at TLCC. Note the `tx_context` field of `SecureTLCCRequest` should give TLCC the context awareness to make sure that for a particular context responses to requests are consistent.  Depending on how the consistency is implemented, it might require though additional messages (see comment below definition of `tx_context`) but i think that should be resolved in #402/#435, not in this PR

Reply:
I am fine with leaving the states as is for now. As we currently use TLCC to only provide integrity metadata, I dont see the need to use this "validated" state as a mechnaimsn for ecc access control. That is, that only an ecc instance that "really" belongs to the given chaincode can access the integrity metadata or "any other data". 

Without this access control, I also think that the `validated` state is not needed. In case a new ecc instance starts, and would access tlcc even before the registration is committed, ecc can indeed process a transaction invocation and access state, however, this transaction will never be commited as the ecc instance is not registred, thus, signature verification will fail.

I would not rename to `..._unvalidated_chaincode_id` but added a comment doesnt harm.